### PR TITLE
Provisioning Info Space Topology

### PR DESCRIPTION
**Pull request by manadart**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

This change is one of several intended to enable spaces and multi-NIC support on Openstack.

It adds a new version of the provisioner API facade.

This version changes the implementation of `ProvisionerInfo` so that instead of using only a single space to determine possible deployment subnet/AZ combinations, all positive space constraints are represented.

A new parameter type accommodates the extra detail.

## QA steps

These changes will materialise in later patches. For now, unit tests provide validation.

For regression:
- Bootstrap to AWS and add a space
- `juju add-machine --constraints spaces=new-space`

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Provisioning Info Space Topology](https://trello.com/c/qzBOdKck/2288-provisioning-info-space-topology)

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 1]
```diff
@@ -191,6 +191,13 @@ type ProvisionerAPIV8 struct {
 // ProvisionerAPIV9 provides v9 of the provisioner facade.
 // Added SupportedContainers
 type ProvisionerAPIV9 struct {
+	*ProvisionerAPIV10
+}
+
+// ProvisionerAPIV9 provides v9 of the provisioner facade.
```
Comment (SimonRichardson) [inline apiserver/facades/agent/provisioner/provisioner.go]: You mean version 10 here?

File: apiserver/facades/agent/provisioner/provisioninginfo.go [hunk 2]
```diff
@@ -57,67 +58,115 @@ func (api *ProvisionerAPI) ProvisioningInfo(args params.Entities) (params.Provis
...
 }
 
-func (api *ProvisionerAPI) getProvisioningInfo(m *state.Machine, env environs.Environ) (*params.ProvisioningInfo, error) {
-	cons, err := m.Constraints()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
+func (api *ProvisionerAPI) getProvisioningInfo(m *state.Machine, env environs.Environ) (*params.ProvisioningInfoV10, error) {
+	var err error
+	result := params.ProvisioningInfoV10{}
```
Comment (SimonRichardson) [inline apiserver/facades/agent/provisioner/provisioninginfo.go]: nit: `var result params.ProvisioningInfoV10` (it makes it very obvious that it's not got anything in it just from scan reading)

File: apiserver/facades/agent/provisioner/provisioninginfo.go [hunk 3]
```diff
@@ -57,67 +58,115 @@ func (api *ProvisionerAPI) ProvisioningInfo(args params.Entities) (params.Provis
...
+
+	if result.ImageMetadata, err = api.availableImageMetadata(m, env); err != nil {
+		return result, errors.Annotate(err, "cannot get available image metadata")
+	}
+
+	if result.ControllerConfig, err = api.st.ControllerConfig(); err != nil {
+		return result, errors.Annotate(err, "cannot get controller configuration")
+	}
+
+	var jobs = m.Jobs()
```
Comment (SimonRichardson) [inline apiserver/facades/agent/provisioner/provisioninginfo.go]: nit: `jobs := m.Jobs()`

File: apiserver/facades/agent/provisioner/provisioninginfo.go [hunk 4]
```diff
@@ -248,32 +298,73 @@ func (api *ProvisionerAPI) machineTags(m *state.Machine, jobs []model.MachineJob
 	return machineTags, nil
 }
 
-// machineSubnetsAndZones returns a map of subnet provider-specific id
-// to list of availability zone names for that subnet. The result can
-// be empty if there are no spaces constraints specified for the
-// machine, or there's an error fetching them.
+func (api *ProvisionerAPI) machineSpaceTopology(m *state.Machine) (params.ProvisioningNetworkTopology, error) {
+	topology := params.ProvisioningNetworkTopology{}
```
Comment (SimonRichardson) [inline apiserver/facades/agent/provisioner/provisioninginfo.go]: nit: `var topology ...`

File: apiserver/facades/agent/provisioner/provisioninginfo.go [hunk 5]
```diff
@@ -248,32 +298,73 @@ func (api *ProvisionerAPI) machineTags(m *state.Machine, jobs []model.MachineJob
...
+		subnetsAndZones, err := api.subnetsAndZonesForSpace(m.Id(), spaceName)
+		if err != nil {
+			return topology, errors.Trace(err)
+		}
+
+		// Record each subnet provider ID as being in the space,
+		// and add the zone mappings to our map
+		subnetIDs := make([]string, 0, len(subnetsAndZones))
+		for sID, zones := range subnetsAndZones {
+			// We do not expect unique provider subnets to be in more than one
```
Comment (SimonRichardson) [inline apiserver/facades/agent/provisioner/provisioninginfo.go]: > We do not expect

but will probably happen, can we guard against this?

```go
if _, ok := topology.SubnetAZs[sID]; ok {
      logger.Debug("oh man, didn't expect this to happen")
}
```

File: apiserver/params/internal.go [hunk 6]
```diff
@@ -743,34 +743,80 @@ type AgentVersionResult struct {
...
 	CloudInitUserData map[string]interface{}   `json:"cloudinit-userdata,omitempty"`
 	CharmLXDProfiles  []string                 `json:"charm-lxd-profiles,omitempty"`
 }
 
-// ProvisioningInfoResult holds machine provisioning info or an error.
+// ProvisioningInfo holds machine provisioning info returned by
+// versions 9 and lower of the provisioner API facade.
+type ProvisioningInfo struct {
+	ProvisioningInfoBase
+	SubnetsToZones map[string][]string `json:"subnets-to-zones,omitempty"`
```
Comment (SimonRichardson) [inline apiserver/params/internal.go]: :+1:

File: worker/provisioner/provisioner_task.go [hunk 7]
```diff
@@ -1098,11 +1103,6 @@ func (task *provisionerTask) startMachine(
 		return err
 	}
 
-	// TODO (jam): 2017-01-19 Should we be setting this earlier in the cycle?
-	if err := machine.SetInstanceStatus(status.Provisioning, "starting", nil); err != nil {
```
Comment (SimonRichardson) [inline worker/provisioner/provisioner_task.go]: Driveby or actually a fix?

Comment (manadart) [reply worker/provisioner/provisioner_task.go]: Drive-by; just moved it up as the first action when adding a machine.

**Comment by manadart** [comment]
$$merge$$
